pkg/applicationserver/applicationserver.go

```diff
@@ -588,6 +592,202 @@ func (as *ApplicationServer) handleJoinAccept(ctx context.Context, ids ttnpb.End
+	return as.changeDownlinkQueue(ctx, dev, link, func(ctx context.Context, dev *ttnpb.EndDevice) ([]*ttnpb.ApplicationDownlink, []*ttnpb.ApplicationDownlink, error) {
+		var previousPendingQueue, currentQueue []*ttnpb.ApplicationDownlink
+		unmatched := invalidatedDownlinks
+		if previousPendingSession != nil {
+			previousPendingQueue, unmatched = ttnpb.PartitionDownlinksBySessionKeyIDEquality(previousPendingSession.SessionKeyID, unmatched...)
+		}
+		if dev.Session != nil {
+			currentQueue, unmatched = ttnpb.PartitionDownlinksBySessionKeyIDEquality(dev.Session.SessionKeyID, unmatched...)
+		}
+		logger := log.FromContext(ctx)
+		for _, item := range unmatched {
+			logger.WithFields(log.Fields(
+				"f_port", item.FPort,
+				"f_cnt", item.FCnt,
+				"session_key_id", item.SessionKeyID,
+			)).Warn("Drop downlink message; unknown session key ID")
+			registerDropDownlink(ctx, dev.EndDeviceIdentifiers, item, errUnknownSession)
+		}
+
+		var sourceQueue []*ttnpb.ApplicationDownlink
+		var sourceSession *ttnpb.Session
+		switch {
+		case previousPendingSession == nil && dev.Session == nil:
+			// We cannot decode any of the downlinks if both sessions are missing.
+			return nil, nil, errNoDeviceSession
+		case dev.Session != nil:
+			// In the presence of a current session we copy the items from the current session to the pending session.
+			sourceQueue = currentQueue
+			sourceSession = dev.Session
+			logger.Debug("Initializing pending downlink queue from the current session")
+		case previousPendingSession != nil:
+			// In the presence of a previous session and the absence of a current session we migrate the items from the
+			// previous pending session to the new pending session.
+			sourceQueue = previousPendingQueue
+			sourceSession = previousPendingSession
+			logger.Debug("Initializing pending downlink queue from the previous pending session")
+		}
+
+		newPendingQueue, err := as.migrateDownlinkQueue(ctx, dev.EndDeviceIdentifiers, sourceQueue, sourceSession, dev.PendingSession, 1)
+		if err != nil {
+			return nil, nil, err
+		}
+		return currentQueue, newPendingQueue, nil
+	})
+}
+
+// replaceDownlinkQueue replaces the current and pending queues of the end device with the provided downlinks.
+// This method requires the given end device's session to be set. This method mutates the end device's session LastAFCntDown.
+// This method mutates the end device's pending session LastAFCntDown if the pending session exists and replacePending is true.
+// This method does not change the contents of the given downlinks.
+func (as *ApplicationServer) replaceDownlinkQueue(ctx context.Context, dev *ttnpb.EndDevice, link *link, previousSession *ttnpb.Session, downlinks []*ttnpb.ApplicationDownlink, nextAFCntDown uint32, replacePending bool) error {
```

Comment:
I think what this method should do, is only recalculation of the active queue with a higher FCnt value. It should never modify the pending queue, because that is never necessary. You don't need the `previousSession` either (it's always `dev.Session`)
Also, name it `recalculateDownlinkQueue` like it was before.